### PR TITLE
[FLINK-20894][Table SQL / API] Introduce SupportsAggregatePushDown interface

### PR DESCRIPTION
**Pull request by sebastianliu**
## What is the purpose of the change

*This pull request introduces a new TableSourceAbility interface for supporting aggregation push down.*

## Brief change log

  - *A new interface of SupportsAggregatePushDown*
  - *A new implementation of ResolvedExpression which is AggregateExpression*
  - *Implement the new SupportsAggregatePushDown interface for TestValuesScanTableSource*

## Verifying this change

This change is only introduced a few new interfaces, and implemented for a test class, which already covered by existing tests, such as *TableSourceITCase*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * <p>A {@link ScanTableSource} extending this interface is able to aggregate records before
+ * returning.
+ */
+@PublicEvolving
+public interface SupportsAggregatePushDown {
+
+    /**
+     * Check and pick all aggregate expressions this table source can support. The passed in
+     * aggregates and grouping sets have been keep in original order.
+     *
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java:61]: Mention how to construct the output row, that the grouping keys should come first, and then the aggregates result. And give an example for better understanding.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+     * <p>Note: The final output data type emitted by the source changes from the original produced
+     * data type to the emitted data type of local aggregate. The passed {@code DataType
+     * producedDataType} is the updated data type for convenience.
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java:64]:
```suggestion
     * <p>Note: Use the passed data type instead of {@link TableSchema#toPhysicalRowDataType()} for
     * describing the final output data type when creating {@link TypeInformation}. The projection of grouping keys and aggregate values is already considered in the
     * given output data type.
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ *
+ * <p>A {@link ScanTableSource} extending this interface is able to aggregate records before
+ * returning.
+ */
+@PublicEvolving
+public interface SupportsAggregatePushDown {
+
+    /**
+     * Check and pick all aggregate expressions this table source can support. The passed in
+     * aggregates and grouping sets have been keep in original order.
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java:60]:
```suggestion
     * Provides a list of aggregate expressions and the grouping keys. The source should pick all the aggregates or nothing and return whether all the aggregates have been pushed down into the source.
```

**Review by wuchong — COMMENTED** [review]
Thanks @sebastianliu , I left some comments.

**Comment by sebastianliu** [comment]
Hi @wuchong , appreciate for your time and thx a lot for your carefully review. I have resolved all of comments. Looking forward another round code review when you have time.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+ *    +- Exchange(distribution=[hash[name, type]])
+ *       +- LocalHashAggregate(groupBy=[name, type], select=[name, type, Partial_SUM(amount) AS sum$0, Partial_MAX(price) AS max$1, Partial_AVG(price) AS (sum$2, count$3), Partial_COUNT(*) AS count1$4])
+ *          +- TableSourceScan(table=[[inventory, project=[name, type, amount, price]]], fields=[name, type, amount, price])
+ * }</pre>
+ *
+ * <p>For efficiency, a source can push local aggregates further down in order to reduce the network
+ * and computing overhead. The passed aggregate functions and grouping sets are in the original
+ * order. The downstream storage which has aggregation capability can directly return the aggregated
+ * values if the underlying database or storage system has aggregation capability. The optimized
+ * plan will change to the following pattern with local aggregate push down:
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java:76]:
```suggestion
 * <p>For efficiency, a source can push local aggregates further down into underlying database or storage system to reduce the network
 * and computing overhead. The passed aggregate functions and grouping sets are in the order defined by the query.
 * The source can directly return the aggregated
 * values if the underlying database or storage system has aggregation capability. The optimized
 * plan will be changed to the following pattern with local aggregate push down:
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+ * <p>Due to the complexity of aggregate, only limited aggregate functions are supported at present.
+ *
+ * <ul>
+ *   <li>Only support sum/min/max/count/avg(will convert to sum0 + count) aggregate function push
+ *       down.
+ *   <li>Only support simple group type, cube and roll up will not be pushed down.
+ *   <li>If expression is involved in aggregate or group by, e.g. max (col1 + col2) or group by
+ *       (col1 + col2), aggregate will not be pushed down.
+ *   <li>Window aggregate function will not be pushed down.
+ *   <li>Aggregate function with filter will not be pushed down.
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java:98]: Do we need to have these limitation in the interface/planner? From my point of view, most of the limitation is defined by connectors. For example, some databases can support grouping set and agg filter, we can push everything we have and make the connector to decide whether they can be pushed down.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+ *          +- Calc(select=[name, type, price, EXPR$0, EXPR$1, EXPR$2, =(CASE(=($e, 0:BIGINT), 0:BIGINT, 1:BIGINT), 0) AS $g_0, =(CASE(=($e, 0:BIGINT), 0:BIGINT, 1:BIGINT), 1) AS $g_1])
+ *             +- HashAggregate(groupBy=[name, type, price, $e], select=[name, type, price, $e, Final_SUM(sum$0) AS EXPR$0, Final_MAX(max$1) AS EXPR$1, Final_AVG(sum$2, count$3) AS EXPR$2])
+ *                +- Exchange(distribution=[hash[name, type, price, $e]])
+ *                   +- LocalHashAggregate(groupBy=[name, type, price, $e], select=[name, type, price, $e, Partial_SUM(amount) AS sum$0, Partial_MAX(price_0) AS max$1, Partial_AVG(price_0) AS (sum$2, count$3)])
+ *                      +- Expand(projects=[name, type, amount, price, $e, price_0], projects=[{name, type, amount, price, 0 AS $e, price AS price_0}, {name, type, amount, null AS price, 1 AS $e, price AS price_0}])
+ *                         +- TableSourceScan(table=[[inventory, project=[name, type, amount, price]]], fields=[name, type, amount, price])
+ * }</pre>
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java:130]: I also think this is not a good example, and we don't need to mention it. Some databases can support count(distinct)

**Review by wuchong — COMMENTED** [review]
Thanks for the updating @sebastianliu , the interface and javadoc looks better now. It would be great if @twalthr could help to have a look too. 

Besides, I just find that we didn't include order expression in `AggregateExpression`, i.e. the `WITHIN GROUP` clause in SQL, Calcite also defines it in `org.apache.calcite.rel.core.AggregateCall#getCollation`.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java (hunk 5 above)
Comment (sebastianliu) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java]: I have updated the limitation part. In the new limitation, we don't restrict the function type, but we do add some restrictions on the complex aggregate patterns.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java (hunk 6 above)
Comment (sebastianliu) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java]: Have fixed, change to use aggregate with filter as an unable push down demo.

**Comment by sebastianliu** [comment]
Thx a lot for the review and suggestion @wuchong , I have refined the javadoc and updated some limitations. For the `Collation` semantic, as we haven't passed the relevant conditions in the `WITHIN GROUP` clause to the aggregate function at present. I think the `AggregateExpression` could exclude the `OrderExpression` for the moment. After the aggregate function can get the order conditions, I can design the `OrderExpression` and add it back.

Look forward for the further progress.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+ * <p>Note: currently, the {@link SupportsAggregatePushDown} is not supported by planner.
+ */
+@PublicEvolving
+public interface SupportsAggregatePushDown {
+
+    /**
+     * Provides a list of aggregate expressions and the grouping keys. The source should pick all
+     * the aggregates or nothing and return whether all the aggregates have been pushed down into
+     * the source.
+     *
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java:137]:
```
In the example mentioned in {@link SupportsAggregatePushDown}, this method would receive groupingSets {code List([1, 4])} which is equivalent to {@code List(["name", "type"])}.

```

File: flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/factories/TestValuesTableFactory.java [hunk 8]
```diff
@@ -619,7 +630,8 @@ private ChangelogMode parseChangelogMode(String string) {
                     SupportsFilterPushDown,
                     SupportsLimitPushDown,
                     SupportsPartitionPushDown,
-                    SupportsReadingMetadata {
+                    SupportsReadingMetadata,
+                    SupportsAggregatePushDown {
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/factories/TestValuesTableFactory.java:634]: Could you revert the changes on this class? Because it's hard to know whether it is implemented correctly before the planner supports `SupportsAggregatePushDown`.

**Comment by sebastianliu** [comment]
Thx a lot for the follow-up review @wuchong , I have resolved comments. Looking forward for the further progress.

**Review by wuchong — CHANGES_REQUESTED** [review]
Thanks for the great work @sebastianliu , LGTM.

However, the code format is failed, could you fix that? 

https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=12808&view=logs&j=52b61abe-a3cc-5bde-cc35-1bbe89bb7df5&t=54421a62-0c80-5aad-3319-094ff69180bb

**Comment by sebastianliu** [comment]
> Thanks for the great work @sebastianliu , LGTM.
> 
> However, the code format is failed, could you fix that?
> 
> https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=12808&view=logs&j=52b61abe-a3cc-5bde-cc35-1bbe89bb7df5&t=54421a62-0c80-5aad-3319-094ff69180bb

Hi @wuchong , sorry for the code style error. I have fixed the format error and tested in my local env. I'll pay more attention later and hope you can help to review another round. Thx again for the patient review.

**Comment by wuchong** [comment]
@sebastianliu never mind, you can setup the code formatting tool in your IDEA, this will can help you automatically format your code. https://ci.apache.org/projects/flink/flink-docs-master/flinkDev/ide_setup.html#code-formatting

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java [hunk 9]
```diff
@@ -135,15 +135,16 @@
...
+     * {@link TypeInformation}. The projection of grouping keys and aggregate values is already
+     * considered in the given output data type. The passed data type pattern is {@code grouping
+     * sets} + {@code aggregate function result}, downstream storage need to organize the returned
+     * aggregate data strictly in this manner.
      *
-     * @param groupingSets a array list of the grouping sets. In the example mentioned in
-     *     {@link SupportsAggregatePushDown}, this method would receive the groupingSets of
-     *     {code List([1, 4]) which is equivalent to {@code List(["name", "type"])}.
+     * @param groupingSets a array list of the grouping sets. In the example mentioned in {@link
+     *     SupportsAggregatePushDown}, this method would receive the groupingSets of {code List([1,
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java:146]: `{code List([1` ==> `{@code List([1`

Comment (sebastianliu) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsAggregatePushDown.java]: fixed, sorry for this typo

**Review by wuchong — APPROVED** [review]
LGTM.

**Comment by wuchong** [comment]
The failed case `UnalignedCheckpointITCase.execute` is not related to this issue. 

Will merge the PR.
